feat: add from_file

svmnotn opened the pull request:
Because of files I had to remove the lifetime requirement on `Source`. Honestly this is the only time that I wish rust was any other lang... all this for some simple strings.... arrg. Oh well....

svmnotn committed:
feat: add from_file

svmnotn commented:
On the note of adding things to `Source` we probably want a `to_ast` that will do the parsing. Or we could have all that as convenience methods on `lib.rs`. Either way I would like a `ast_from_file` and `ast_from_text` method in graphql-language.

calebmer commented on graphql-parser/src/source.rs:
```diff
@@ -1,23 +1,46 @@
-impl<'a> Source<'a> {
-  pub fn new(body: &'a str) -> Source<'a> {
+impl Source {
+  pub fn new(body: String) -> Self {
```
How about making the signature `pub fn new(name: Option<String>, body: String)`? Or use the setter patter so: `Source::new(body).set_name(name)`. I'd prefer the option approach.

calebmer commented on graphql-parser/src/source.rs:
```diff
@@ -1,23 +1,46 @@
-}
\ No newline at end of file
+
+  pub fn from_file<P: AsRef<Path>>(path: P) -> Result<Self, io::Error> {
```
We should really create our own error type and have it impl `From<io::Error>`.

cksac commented:
I would like to keep it as &'a str to avoid extra allocation as most of the time query will not from files.

calebmer commented on graphql-parser/src/source.rs:
```diff
@@ -1,23 +1,46 @@
+    Ok(Source::with_name(Self::get_file_name(p), s))
+  }
+
+  fn get_file_name(path: PathBuf) -> String {
```
This probably shouldn't be in the `impl` block, in fact this probably shouldn't be a function at all. Here it is with functional programming:

``` rust
path.file_name().and_then(OsStr::to_str).unwrap_or(DEFAULT_SOURCE_NAME).to_owned()
```

And if that line is a little too long for your taste:

``` rust
let name = {
  path
  .file_name()
  .and_then(OsStr::to_str)
  .unwrap_or(DEFAULT_SOURCE_NAME)
  .to_owned()
};
```

Much easier to understand then the `if let Some/Ok`s.

svmnotn committed:
feat: module level `from_file`

Making it at the module level solves many of the lifetime issues.
Also made Source &'a str again, and lowered the imports

svmnotn commented:
review?
